pkts-core/src/main/java/io/pkts/packet/sip/impl/SipRequestPacketImpl.java

```diff
@@ -49,9 +49,9 @@ public SipRequestPacket toRequest() throws ClassCastException {
      */
     @Override
     public SipRequestPacket clone() {
-        final TransportPacket transport = getTransportPacket().clone();
+        final TransportPacket transport = getTransportPacket();
         final SipRequest request = (SipRequest) getSipMessage().clone();
-        return new SipRequestPacketImpl(transport, request);
+        return new SipRequestPacketImpl(transport != null ? transport.clone() : null, request);
```

Comment:
under what circumstances would the `TransportPacket` be null? It really shouldn't so curious to why you removed that check.

Reply:
For flat SIP messages (plain text) that could be sent through kafka topics for instance:
https://www.kamailio.org/docs/modules/5.1.x/modules/dmq.html (not kafka :) )

Reply:
yeah, then you shouldn't use the `SipPacket` you should only use the `SipMessage` which is what you would use if you were to build your own sip stack etc. This is ONLY for pcaps... so, you should just use this: https://github.com/aboutsip/pkts/tree/master/pkts-sip and you see examples on siplib.io and also sipstack.io.

Reply:
I believe the way to go is to use the Framer API. Since both PCAP and SIP has it, why not?
Check this out: https://github.com/apache/metron.
Another project reading pcaps from kafka (previous one was reading flat from dmq).
"Framers" should be the access point to an agnostic entry point. That's what APIs are for.
Why should I repeat your code inside SipFramer?

Reply:
The Framers are different. Every protocol has a framer. I have a framer for Diameter, one for GTP, one for SIP and one for PCAPs. You use the appropriate framer for the situation. You would use the PCAP framers to frame pcaps and once you get out the "raw" bytes for a particular protocol, you would use that framer. 

So, the `xxxPacket` are only to be used when you are dealing with pcaps. If you have a SIP message as a `String`, which you may have gotten from a text file, or if you extracted that by reading from a network socket (and really, then you got `byte[]`) then why wrap that in a pcap encapsulation? So, for any protocol, all you have to do is `<protocol_framer/parser>.frame(str)`.

Another example: if I deal with HTTP and I read that off a TCP socket, I would only want to frame that HTTP message. It is not coming in over the socket as "pcap" so why wouldn't I just do `Http.frame(byte[])`?

So again, the entire purpose of `SipRequestPacket` is to represent, or extend, a `SipRequest` that is stored in a `pcap`. If it is not stored in a pcap then you should just use `SipRequest` etc.

Reply:
Framer API would work for most cases whether you are using PCAP or Raw SIP messages.
In the worst case scenario you can even remove parent limitation from IPv4, TCP, UDP, etc...
Don't see how enforcing that condition would make it more robust.

> Every protocol has a framer
If so, PCAP works independently and so, SIP also should.

Using static methods makes it difficult to write an abstraction.
Our fraud system should be able to read both (for now). I'm already able to, just trying to make this more robust and agnostic.
But, your project your architecture :)
Happy to contribute since we'll be using this a lot from now on.

Reply:
yeah, if you could share more about the use case because, again, the framers that is trying to figure out if is protocol A, B or C is meant only for pcaps and then they actually call e.g. `SipParser.frame`. So, currently, I will not relax the rules around `xxxPackets` because their sole purpose is that it's a "packet" from a pcap (packet capture) and as such, it MUST have a parent frame all the way up until the actual Pcap frame itself.

You could just as easily do a switch yourself in order to determine if:
1. is it a pcap?
2. Is it a raw SIP message
3. is it a raw HTTP message

etc etc. If the parent packet is now optional, you should also change the signature to indicate that it is indeed `Optional`. In a pcap, it is not.

Reply:
I already implemented the solution. It's working for both pcap and raw sip message.
Frame API is well thought and it shouldn't be bound to pcap.
For instance if I use tcpdump/ngrep/sipgrep and send the output directly to kafka. What framers of those already developed would you reuse?
The one I think it's global to all is SIPFramer.

Reply:
Should I close this?
I need at least the method for getting all headers.
Also I will propose changing to a LinkedHashMap (if order matters) because we have 3 different projects in Europe and APAC that required a few more headers like User-Agent, Timestamp, Diversion, etc...
Thanks

Reply:
sorry, got busy with work.

Yeah, I still don't see the need and it breaks my intent of the "packet" within a pcap. All you accomplish is create a very thin wrapper around the SipMessage but you don't get anything for it. After all, the "only" benefit from the `packet` structure is that you do have e.g. the tcp frame, ip frame etc etc but now that you null that out, you only get a thin wrapper and nothing else.

You request to getting all the headers is perfectly fine though.

Also, the LinkedHashMap - if you can prove that this is vastly faster and as such, actually matters, sure. I'm not so sure it is. I assume speed is your main concern? If it is, again, I would say do your performance test first to validate that speed is an issue...